apollo-ios-codegen/Sources/GraphQLCompiler/JavaScript/src/utilities/typePolicyResolver.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  DirectiveNode,
+  GraphQLCompositeType,
+  GraphQLError,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLSchema,
+  isInterfaceType,
+  isNonNullType,
+  isScalarType,
+  isUnionType,
+  Kind,
+  valueFromASTUntyped,
+} from "graphql";
+import { directive_typePolicy } from "./apolloCodegenSchemaExtension";
+
+const directiveName = directive_typePolicy.name.value;
+
+type TypePolicyDirectiveResult = {
+  directive: DirectiveNode;
+  source: GraphQLObjectType | GraphQLInterfaceType;
+};
+
+function matchDirectiveArguments(
+  first: DirectiveNode,
+  second: DirectiveNode
+): boolean {
+  return (
+    (first.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString() ===
+    (second.arguments ?? [])
+      .map((node) =>
+        JSON.stringify([node.name.value, valueFromASTUntyped(node.value)])
+      )
+      .sort()
+      .toString()
+  );
+}
+
+function typePolicyDirectiveFor(
+  type: GraphQLCompositeType
+): TypePolicyDirectiveResult | undefined {
+  if (isUnionType(type)) {
+    return undefined;
+  }
+
+  for (const extension of type.extensionASTNodes ?? []) {
+    const directive = extension.directives?.find(
+      (d) => d.name.value === directiveName
+    );
+    if (directive) {
+      return { directive, source: type };
+    }
+  }
+
+  for (const directive of type.astNode?.directives ?? []) {
+    if (directive.name.value === directiveName) {
+      return { directive, source: type };
+    }
+  }
+
+  if ("getInterfaces" in type) {
+    let result: TypePolicyDirectiveResult | undefined;
+
+    for (const interfaceType of type.getInterfaces()) {
+      const found = typePolicyDirectiveFor(interfaceType);
+      if (!found) continue;
+
+      if (!result) {
+        result = found;
+      } else if (!matchDirectiveArguments(result.directive, found.directive)) {
+        throw new GraphQLError(
+          `Type "${type.name}" inherits conflicting @typePolicy directives from interfaces "${result.source.name}" and "${found.source.name}". Please specify a @typePolicy directive on the concrete type.`,
+          { nodes: type.astNode }
```

Comment:
I think letting types override their inherited `@typePolicy` is a good idea, and it should probably be added to Apollo Kotlin as well.

Reply:
I'd challenge that. In the case of interfaces (`Node`) and new objects being added without the client recompiling, this can lead to cache misses (see [this comment](https://github.com/apollographql/apollo-ios-dev/pull/554#discussion_r1874023173) from @AnthonyMDev). 

So we have already a use case for not allowing overriding `@typePolicy`. Not saying we shouldn't do it but it'd need a strong documented use case before we can make an informed tradeoff.

Reply:
@martinbonnin, are you suggesting that you can only have a `@typePolicy` **EITHER** on a type or on a single interface it implements, but not both?

Reply:
Yes, for the same reason as in https://github.com/apollographql/apollo-ios-dev/pull/554#discussion_r1882997857. It's not very flexible but do we need that flexibility?

Reply:
I'm not opposed to disallowing `@typePolicy` overrides for the time being, since it could always be re-added at a later stage without changing existing behavior.